api/src/main/java/org/pmiops/workbench/db/dao/BillingProjectBufferEntryDao.java

```diff
@@ -40,6 +44,29 @@
     return DaoUtils.getAttributeToCountMap(findAll(), DbBillingProjectBufferEntry::getStatusEnum);
   }
 
+  @Query(value = "select status, count(billing_project_buffer_entry.billing_project_buffer_entry_id) as num_projects\n"
+      + "    from billing_project_buffer_entry\n"
+      + "group by status\n"
+      + "order by status;", nativeQuery = true)
+//  List<BillingProjectBufferEntryStatusToCountResult> computeProjectCountByStatus();
+   BillingProjectBufferEntryStatusToCountResult[] computeProjectCountByStatus();
```

Comment:
Driving by but is it necessary to add this intermediate object? I think the interface we had, `getCountByStatusMap()`, works well. We just need to change the implementation to speed it up.

The `computeProjectCountByStatus2()` call below should work and that gets us most of the way there. All `getCountByStatusMap()` would have to do is convert the result from `computeProjectCountByStatus2()` to match the existing interface.

Reply:
I don't know how to capture the query result safely without a new POJO or entity. Like, that's what I'm supposed to get from. packages. like this I thought. If I'm having to. cast from. object and then. manually call a constructor with parameters the same names as my columns, then. I want to shop for another framework.  Ideally I could just make an Immutable for it.

But if that works, it should help.

Reply:
@jaycarlton If you would like to catch the query output in an object but don't want to have to annotate another entity... you can use a Projection interface. It allows you to project a subset of properties from your dao into another object that isn't your actual entity.
Projection interface example here: https://github.com/all-of-us/workbench/blob/freemabd/address-comments/common-api/src/main/java/org/pmiops/workbench/cdr/model/DbMenuOption.java#L3-L3 
Usage example here: https://github.com/all-of-us/workbench/blob/freemabd/address-comments/common-api/src/main/java/org/pmiops/workbench/cdr/dao/CBCriteriaDao.java#L238-L238

Reply:
oo, this is neat! thanks @freemabd 

Also, another comment, I think its best to avoid using native queries where possible. Native queries will run the query as is against the underlying database. This would be OK if we used the same database everywhere but I believe our tests use some in memory database (H2?).

I have had issues where native queries work in production and fail in tests and vice versa.

Writing HQL will let hibernate figure out the right way to generate the query based on what database is currently in use. Caveat is that you don't have access to all the SQL tools that MySQL provides but HQL should suffice for most cases.

Reply:
Great points. Was going to try going up a level. in abstraction after getting something working.

Reply:
Projection interface works! Much nicer. No annotatitons even.